packages/issuer/lib/VcIssuer.ts

```diff
@@ -82,7 +85,33 @@ export class VcIssuer {
     throw new Error(TokenErrorResponse.invalid_request)
   }
 
-  private async validateJWT(issueCredentialRequest: CredentialRequest, jwtVerifyCallback?: JWTVerifyCallback, clientId?: string, preAuthorizedCode?: string): Promise<void> {
+  private async retrieveGrantsFromClient(issuerState: string): Promise<{ clientId: string; grants: Grant }> {
+    const credentialOfferState: CredentialOfferState | undefined = await this.retrieveCredentialOffer(issuerState)
+    const clientId = credentialOfferState?.clientId
+    const grants = credentialOfferState?.credentialOffer?.grants
+    if (!clientId) {
+      throw new Error(UNDEFINED_CLIENT_ID)
+    }
+    if (!grants?.authorization_code?.issuer_state || !grants['urn:ietf:params:oauth:grant-type:pre-authorized_code']?.['pre-authorized_code']) {
+      throw new Error(GRANTS_MUST_NOT_BE_UNDEFINED)
+    }
+    return { clientId, grants }
+  }
+
+  private async retrieveCredentialOffer(issuerState: string): Promise<CredentialOfferState | undefined> {
+    if (await this._stateManager.hasState(issuerState)) {
+      return await this._stateManager.getState(issuerState)
+    } else {
+      throw new Error(UNKNOWN_CLIENT_ERROR)
```

Comment:
IMO this should be a method on the stateManager, something like getAssertedState

Reply:
Fixed!